Custom Translation Feature - Show Translation

JoshuaSchaefer opened the pull request:
Automatically show a custom translation for the program description and/or credits to the user based on the user's preferred language. If there is not a custom translation defined for the user's language, then just show the default description and/or credits.

---
### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroweb-Symfony/blob/develop/.github/contributing.md) and [wiki pages](https://github.com/Catrobat/catroweb-Symfony/wiki/) of this repository.

- [ ] Include the name and id of the Jira ticket in the PR’s title eg.: `SHARE-666 The devils ticket`
- [ ] Choose the proper base branch (*develop*)
- [ ] Confirm that the changes follow the project’s coding guidelines
- [ ] Verify that the changes generate no warnings and errors 
- [ ] Verify to commit no other files than the intentionally changed ones
- [ ] Include reasonable and readable tests verifying the added or changed behavior
- [ ] Verify that all tests are passing (CI), if not please state the test cases in the [section](#Tests) below
- [ ] Perform a self-review of the changes
- [ ] Stick to the project’s git workflow (rebase and squash your commits)
- [ ] Verify that your changes do not have any conflicts with the base branch
- [ ] Put your ticket into the `Code Review` section in [Jira](https://jira.catrob.at/)
- [ ] Post a message in the *#catroweb* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer
- [ ] Check that your pull request has been successfully deployed to https://web-test-1.catrob.at/

### Additional Description
If there is a custom translation defined for the language the user has selected in the footer, the translation is automatically shown in the description and/or credits when the page loads. However, the custom translation is not shown if the user is the program owner.

### Tests - additional information
There are new tests that verify that the custom translation is automatically shown if there is a translation matching the user's language and does not show if there is not a matching translation.

danielzhang130 reviewed (commented):
i'm slightly concerned with the parsing cookie code here and in the old machine translation code. I think we can pass the argument more cleanly from the backend - not that it has to be done in this pr

JoshuaSchaefer commented:
> i'm slightly concerned with the parsing cookie code here and in the old machine translation code. I think we can pass the argument more cleanly from the backend - not that it has to be done in this pr

Yeah, we can create a backend task to create a new API call to get the current language and a frontend task to use it in the show custom translation and machine translation code

danielzhang130 commented:
rebase with the latest develop so it's easier to review. Also if you're ok with the changes with cookies in the other pr, we can do the same here

danielzhang130 commented on assets/js/custom/ProgramSection.js:
```diff
@@ -0,0 +1,28 @@
+    customTranslationApi.getCustomTranslation(
+      this.programId,
+      usersLanugage,
+      callback
```
Is this implying a no-op error callback so when there are no defined translations, it will do nothing?

JoshuaSchaefer replied:
Yeah, it just won't do anything if there is an error because most of the time it just returns a 404 not found if there isn't any custom translation found for the language

danielzhang130 replied:
That's a bit too implicit for me.  Can we change the name of this argument to notAvailableCallback

https://github.com/danielzhang130/Catroweb/blob/9349d70a6763ecb1324e096f407b3bbf2a71849d/assets/js/api/CustomTranslationApi.js#L8

and pass a empty function here, or some other way to make it clearer that this function doesn't ways replace with custom translation?

```
noop = () => {}
```

```suggestion
      callback,
      noop
```

JoshuaSchaefer replied:
Added a noop function as a callback

danielzhang130 commented on tests/behat/features/web/translation/show_custom_translation.feature:
```diff
@@ -0,0 +1,37 @@
+    Then the selected language should be "English"
+    And I wait for the page to be loaded
+    And I should see "my description"
+    But I should not see "translated description"
```
I think we should use a separate scenario for "not showing custom translation". Also I think we could merge the two scenarios on "description" and "credit" into one scenario. So all in all, one scenario for showing custom translation, one for not showing any.

JoshuaSchaefer replied:
Update to two tests, one showing a translation and one not showing the translation

danielzhang130 commented:
> I was thinking that we could eventually expand the ProgramSection file by adding some of the shared code between the ProgramDescription and ProgramCreadits files. 

Does your vision of the `ProgramSection` class include the editor? I was planning to extract the editor functionality into its own class. https://capstone-pm.atlassian.net/browse/SE-42

JoshuaSchaefer commented:
> > I was thinking that we could eventually expand the ProgramSection file by adding some of the shared code between the ProgramDescription and ProgramCreadits files.
> 
> Does your vision of the `ProgramSection` class include the editor? I was planning to extract the editor functionality into its own class. https://capstone-pm.atlassian.net/browse/SE-42

Yeah, I was thinking we could move some of the repeated code for the editor in this class. Most of the editor functionality should probably be separate since it has to do with specific elements, right? I was mainly thinking we could move the saveDescription/saveCredits function into this class. We could create another class/function for the custom translation editor though if you think that is the best option

danielzhang130 commented on assets/js/custom/ProgramCredits.js:
```diff
@@ -24,6 +24,8 @@ export function ProgramCredits (programId, defaultText, translationSaved, transl
     if (myProgram) {
       getLanguages()
+    } else {
+      programSection.showCustomTranslation(customTranslationApi, setCredits)
```
Does this work?

```suggestion
      programSection.showCustomTranslation(customTranslationApi, credits.text)
```

JoshuaSchaefer replied:
Based on some testing it doesn't look like it works. Good idea though, would have been nice if it worked

danielzhang130 commented on assets/js/custom/ProgramSection.js:
```diff
@@ -0,0 +1,28 @@
+export class ProgramSection {
```
I think we should rename this to be more specific to showing custom translation, or even merge this into CustomTranslationApi, once we don't have the cookie parsing chunk

JoshuaSchaefer replied:
I ended up removing the class since it was more code than just calling the API class directly. We can add something like this when we want to merge some of the repeated code in the sections

Commit:
Custom Translation Feature - Show Translation

Automatically show a custom translation for the program description and/or credits to the user based on the user's preferred language. If there is not a custom translation defined for the user's language, then just show the default description and/or credits.
